Code/Framework/AzCore/AzCore/DOM/DomPath.cpp

```diff
@@ -0,0 +1,422 @@
+    }
+
+    PathEntry& PathEntry::operator=(AZ::Name value)
+    {
+        m_value = AZStd::move(value);
+        return *this;
+    }
+
+    PathEntry& PathEntry::operator=(AZStd::string_view value)
+    {
+        m_value = AZ::Name(value);
+        return *this;
+    }
+
+    bool PathEntry::operator==(const PathEntry& other) const
+    {
+        return m_value == other.m_value;
+    }
+
+    bool PathEntry::operator==(size_t value) const
+    {
+        return IsIndex() && GetIndex() == value;
+    }
+
+    bool PathEntry::operator==(const AZ::Name& key) const
+    {
+        return IsKey() && GetKey() == key;
+    }
+
+    bool PathEntry::operator==(AZStd::string_view key) const
+    {
+        return IsKey() && GetKey() == AZ::Name(key);
+    }
+
+    bool PathEntry::operator!=(const PathEntry& other) const
+    {
+        return m_value != other.m_value;
+    }
+
+    bool PathEntry::operator!=(size_t value) const
+    {
+        return !IsIndex() || GetIndex() != value;
+    }
+
+    bool PathEntry::operator!=(const AZ::Name& key) const
+    {
+        return !IsKey() || GetKey() != key;
+    }
+
+    bool PathEntry::operator!=(AZStd::string_view key) const
+    {
+        return !IsKey() || GetKey() != AZ::Name(key);
+    }
+
+    void PathEntry::SetEndOfArray()
+    {
+        m_value = EndOfArrayIndex;
+    }
+
+    bool PathEntry::IsEndOfArray() const
+    {
+        return AZStd::holds_alternative<size_t>(m_value) && AZStd::get<size_t>(m_value) == EndOfArrayIndex;
+    }
+
+    bool PathEntry::IsIndex() const
+    {
+        return AZStd::holds_alternative<size_t>(m_value) && AZStd::get<size_t>(m_value) != EndOfArrayIndex;
+    }
+
+    bool PathEntry::IsKey() const
+    {
+        return AZStd::holds_alternative<AZ::Name>(m_value);
+    }
+
+    size_t PathEntry::GetIndex() const
+    {
+        return AZStd::get<size_t>(m_value);
+    }
+
+    const AZ::Name& PathEntry::GetKey() const
+    {
+        return AZStd::get<AZ::Name>(m_value);
+    }
+
+    Path::Path(AZStd::initializer_list<PathEntry> init)
+        : m_entries(init)
+    {
+    }
+
+    Path::Path(AZStd::string_view pathString)
+    {
+        FromString(pathString);
+    }
+
+    Path Path::operator/(const PathEntry& entry) const
+    {
+        Path newPath(*this);
+        newPath.Push(entry);
+        return newPath;
+    }
+
+    Path Path::operator/(size_t index) const
+    {
+        return *this / PathEntry(index);
+    }
+
+    Path Path::operator/(AZ::Name key) const
+    {
+        return *this / PathEntry(key);
+    }
+
+    Path Path::operator/(AZStd::string_view key) const
+    {
+        return *this / PathEntry(key);
+    }
+
+    Path Path::operator+(const Path& other) const
+    {
+        Path newPath(*this);
+        for (const PathEntry& entry : other)
+        {
+            newPath.Push(entry);
+        }
+        return newPath;
+    }
+
+    bool Path::operator==(const Path& other) const
+    {
+        return m_entries == other.m_entries;
```

Comment:
question: Just confirming...Does this result in a operator== check on each PathEntry in the vector?

Reply:
Yep, per the [`std::vector` spec](https://en.cppreference.com/w/cpp/container/vector/operator_cmp) it'll compare length and call the value types' `operator==`. The concatenate test indirectly tests this behavior as well:
```
    TEST_F(DomPathTests, OperatorOverloads_Concatenate)
    {
        EXPECT_EQ(Path("/foo/bar"), Path("/foo") + Path("/bar"));
        EXPECT_EQ(Path("/foo"), Path("/foo") + Path());
        EXPECT_EQ(Path("/foo/1/bar/0"), Path("/foo/1") + Path("/bar/0"));
    }
```